application/application.go

```diff
@@ -0,0 +1,585 @@
+		log.Error(ctx, "State Machine - Publish: PublishEdition: failed to update edition during publishing", err, data)
+		return err
+	}
+
+	return nil
+}
+
+func PublishCMDInstance(smDS *StateMachineDatasetAPI, ctx context.Context,
+	versionUpdate *models.Version, // Called Instances in Mongo
+	data log.Data) error {
+
+	if err := smDS.DataStore.Backend.SetInstanceIsPublished(ctx, versionUpdate.ID); err != nil {
+		if user := dprequest.User(ctx); user != "" {
+			data[reqUser] = user
+		}
+		if caller := dprequest.Caller(ctx); caller != "" {
+			data[reqCaller] = caller
+		}
+		err := errors.WithMessage(err, "State Machine - Publish: PublishInstance: failed to set instance node is_published")
+		log.Error(ctx, "failed to publish instance version", err, data)
+		return err
+	}
+
+	return nil
+
+}
+
+func PublishDataset(smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails,
+	data log.Data) error {
+	// Pass in newVersion variable to include relevant data needed for update on dataset record (e.g. links)
+	if err := smDS.publishDataset(ctx, currentDataset, versionUpdate); err != nil {
+		log.Error(ctx, "State Machine: Publish: PublishDataset: failed to update dataset document once version state changes to publish", err, data)
+		return err
+	}
+	data["type"] = currentVersion.Type
+	data["version_update"] = versionUpdate
+	log.Info(ctx, "State Machine: Publish: PublishDataset: published version", data)
+
+	// Only want to generate downloads again if there is no public link available
+	if currentVersion.Downloads != nil && currentVersion.Downloads.CSV != nil && currentVersion.Downloads.CSV.Public == "" {
+		// Lookup the download generator using the version document type
+		t, err := models.GetDatasetType(currentVersion.Type)
+		if err != nil {
+			return fmt.Errorf("error getting type of version: %w", err)
+		}
+		generator, ok := smDS.DownloadGenerators[t]
+		if !ok {
+			return fmt.Errorf("no downloader available for type %s", t)
+		}
+		// Send Kafka message.  The generator which is used depends on the type defined in VersionDoc.
+		if err := generator.Generate(ctx, versionDetails.datasetID, versionUpdate.ID, versionDetails.edition, versionDetails.version); err != nil {
+			data["instance_id"] = versionUpdate.ID
+			data["state"] = versionUpdate.State
+			data["type"] = t.String()
+			log.Error(ctx, "State Machine: Publish: PublishDataset: error while attempting to generate full dataset version downloads on version publish", err, data)
+			return err
+			// TODO - TECH DEBT - need to add an error event for this.  Kafka message perhaps.
+		}
+		log.Info(ctx, "State Machine: Publish: PublishDataset: generated full dataset version downloads:", data)
```

Comment:
Need to check that generator is not nil. Something like:
```suggestion
		if generator == nil {
			log.Info(ctx, "PublishDataset: no download generation needed for dataset type", data)
		} else {
			if err := generator.Generate(ctx, versionDetails.datasetID, versionUpdate.ID, versionDetails.edition, versionDetails.version); err != nil {
				data["instance_id"] = versionUpdate.ID
				data["state"] = versionUpdate.State
				data["type"] = t.String()
				log.Error(ctx, "State Machine: Publish: PublishDataset: error while attempting to generate full dataset version downloads on version publish", err, data)
				return err
				// TODO - TECH DEBT - need to add an error event for this.  Kafka message perhaps.
			}
			log.Info(ctx, "State Machine: Publish: PublishDataset: generated full dataset version downloads:", data)
		}
```

Reply:
The nil check has now been added